pkg/search/controller.go

```diff
@@ -161,46 +162,219 @@ func (c *Controller) handleErr(err error, key interface{}) {
+	defer func() {
+		updated = cr
+	}()
+	if cr == nil {
+		cr = &clusterRegistry{
+			resources:  make(map[schema.GroupVersionResource]registrySet),
+			registries: make(registrySet)}
+	}
+	for _, name := range added {
+		cr.registries[name] = struct{}{}
+		addedResources = c.getRegistryAddedResources(registries[name], cr, addedResources)
+	}
+	for _, name := range removed {
+		delete(cr.registries, name)
+		for resource, previousRegistries := range cr.resources {
+			delete(previousRegistries, name)
+			if len(previousRegistries) <= 0 {
+				removedResources = append(removedResources, resource)
+			}
+		}
+		for _, resource := range removedResources {
+			delete(cr.resources, resource)
+		}
+	}
+	for _, registry := range matched {
+		addedResources = c.getRegistryAddedResources(registry, cr, addedResources)
+	}
+	return
+}
+
+func (c *Controller) findOpenSearchBackend(matchedRegistries map[string]*searchv1alpha1.ResourceRegistry) *searchv1alpha1.BackendStoreConfig {
+	for _, registry := range matchedRegistries {
+		if backend := registry.Spec.BackendStore; backend != nil && backend.OpenSearch != nil {
+			// one cluster may be related to multi registries,
+			// however only one backend could keep in memory with one cluster
+			return backend
+		}
+	}
+	return nil
+}
 
-	// STEP1: stop informer manager for the cluster which does not exist anymore or is not ready.
-	cls, err := c.clusterLister.Get(cluster)
+func (c *Controller) clusterAbleToCache(cluster string) (cls *clusterv1alpha1.Cluster, able bool, err error) {
+	cls, err = c.clusterLister.Get(cluster)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
 			klog.Infof("try to stop cluster informer %s", cluster)
 			c.InformerManager.Stop(cluster)
-			return nil
+			return
 		}
-		return err
+		return
 	}
 
 	if !cls.DeletionTimestamp.IsZero() {
 		klog.Infof("try to stop cluster informer %s", cluster)
 		c.InformerManager.Stop(cluster)
-		return nil
+		return
 	}
 
 	if !util.IsClusterReady(&cls.Status) {
 		klog.Warningf("cluster %s is notReady try to stop this cluster informer", cluster)
 		c.InformerManager.Stop(cluster)
+		return
+	}
+	able = true
+	return
+}
+
+func (c *Controller) reconcileClusterWithRegistries(cls *clusterv1alpha1.Cluster) (matchedRegistries map[string]*searchv1alpha1.ResourceRegistry, cr *clusterRegistry, resourcesChanged, newRegistry bool, err error) {
+	cluster := cls.Name
+	var allRegistries map[string]*searchv1alpha1.ResourceRegistry
+	var matched []string
+	allRegistries, matched, err = c.getClusterMatchedRegistries(cls)
+	if err != nil {
+		klog.Errorf("get cluster matched resource registries failed, cluster: %s, error: %s", cluster, err)
+		return
+	}
+	if len(matched) > 0 {
+		klog.V(4).Infof("registries matched with cluster, cluster: %s, registries: %s",
```

Comment:
It is recommended to keep the first letter of all log printing in uppercase.

Reply:
OK, these is just for debugging during I code this pr, some of these log seems not necessary, I'm going to refactor logs related to this pr.